### PR TITLE
add release workflows

### PR DESCRIPTION
**Pull request by verbanicm**
**Release Workflows** 

Adds reusable workflows for releasing per new release process discussed.

### CODE REVIEW

File: .github/workflows/draft-release.yml [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+on:
+  workflow_call:
+    inputs:
+      version_strategy:
+        description: 'Version Strategy'
+        required: true
+        type: 'string'
+
+env:
+  PR_BRANCH: 'actions/draft-release-${{ github.ref_name}}'
```
Comment (sethvargo) [inline .github/workflows/draft-release.yml:12]:
```suggestion
  PR_BRANCH: 'actions/draft-release-${{ github.ref_name }}'
```

File: .github/workflows/draft-release.yml [hunk 2]
```diff
@@ -0,0 +1,133 @@
+name: 'Draft Release'
```
Comment (sethvargo) [inline .github/workflows/draft-release.yml:1]:
```suggestion
name: 'Draft release'
```

File: .github/workflows/draft-release.yml [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+
+on:
+  workflow_call:
+    inputs:
+      version_strategy:
+        description: 'Version Strategy'
```
Comment (sethvargo) [inline .github/workflows/draft-release.yml:7]:
```suggestion
        description: 'Version strategy'
```

Can we elaborate on this description? Imagine someone who isn't as familiar with the release process.

File: .github/workflows/draft-release.yml [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+  PR_BRANCH: 'actions/draft-release-${{ github.ref_name}}'
+
+jobs:
+  draft-release:
+    name: 'Draft Release'
+    runs-on: 'ubuntu-latest'
+    steps:
+      - uses: 'actions/checkout@v3'
+
+      - uses: 'actions/setup-node@v2'
```
Comment (sethvargo) [inline .github/workflows/draft-release.yml:21]: v3 is available

File: .github/workflows/draft-release.yml [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+        with:
+          node-version: '16.x'
+
+      - name: 'Build and Push'
+        id: 'build'
+        run: |-
+          git config user.name "github-actions[bot]"
```
Comment (sethvargo) [inline .github/workflows/draft-release.yml:28]: Document why this is needed

File: .github/workflows/draft-release.yml [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+
+          npm version ${{ inputs.version_strategy }} --git-tag-version=false
+
+          NEW_VERSION=$(cat package.json | jq -r .version)
+          echo "::set-output name=new_version::$NEW_VERSION"
+
+          npm ci && npm run build
+
+          git add .
+          git commit -m "Release: v$NEW_VERSION"
```
Comment (sethvargo) [inline .github/workflows/draft-release.yml:41]:
```suggestion
          git commit -m "Release: v${NEW_VERSION}"
```

File: .github/workflows/draft-release.yml [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+        run: |-
+          git config user.name "github-actions[bot]"
+          git config user.email "41898282+github-actions[bot]@users.noreply.github.com"
+
+          git checkout -b $PR_BRANCH
+
+          npm version ${{ inputs.version_strategy }} --git-tag-version=false
+
+          NEW_VERSION=$(cat package.json | jq -r .version)
+          echo "::set-output name=new_version::$NEW_VERSION"
```
Comment (sethvargo) [inline .github/workflows/draft-release.yml:36]:
```suggestion
          echo "::set-output name=new_version::${NEW_VERSION}"
```

File: .github/workflows/draft-release.yml [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+        with:
+          node-version: '16.x'
+
+      - name: 'Build and Push'
+        id: 'build'
+        run: |-
+          git config user.name "github-actions[bot]"
+          git config user.email "41898282+github-actions[bot]@users.noreply.github.com"
+
+          git checkout -b $PR_BRANCH
```
Comment (sethvargo) [inline .github/workflows/draft-release.yml:31]:
```suggestion
          git checkout -b "${PR_BRANCH}"
```

File: .github/workflows/draft-release.yml [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+    name: 'Draft Release'
+    runs-on: 'ubuntu-latest'
+    steps:
+      - uses: 'actions/checkout@v3'
+
+      - uses: 'actions/setup-node@v2'
+        with:
+          node-version: '16.x'
+
+      - name: 'Build and Push'
```
Comment (sethvargo) [inline .github/workflows/draft-release.yml:25]:
```suggestion
      - name: 'Build and push'
```

File: .github/workflows/draft-release.yml [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+          NEW_VERSION=$(cat package.json | jq -r .version)
+          echo "::set-output name=new_version::$NEW_VERSION"
+
+          npm ci && npm run build
+
+          git add .
+          git commit -m "Release: v$NEW_VERSION"
+          git push origin $PR_BRANCH --force
+
+      - name: 'Generate Release Notes'
```
Comment (sethvargo) [inline .github/workflows/draft-release.yml:44]:
```suggestion
      - name: 'Generate release notes'
```

File: .github/workflows/draft-release.yml [hunk 11]
```diff
@@ -0,0 +1,133 @@
...
+            const sha = `${{ github.sha }}`;
+            let previousTagName = ``;
+
+            try{
+              const latestRelease = await github.rest.repos.getLatestRelease({
+                owner: context.repo.owner,
+                repo: context.repo.repo,
+              });
+              previousTagName = latestRelease.data.tag_name;
+            }catch(err){
```
Comment (sethvargo) [inline .github/workflows/draft-release.yml:59]:
```suggestion
            } catch(err) {
```

File: .github/workflows/draft-release.yml [hunk 12]
```diff
@@ -0,0 +1,133 @@
...
+              if(err['status'] !== 404){
```
Comment (sethvargo) [inline .github/workflows/draft-release.yml:60]:
```suggestion
              if(err['status'] !== 404) {
```

File: .github/workflows/draft-release.yml [hunk 13]
```diff
@@ -0,0 +1,133 @@
...
+                core.error(`Failed to load latest release: ${err}`);
+                process.exit(1);
```
Comment (sethvargo) [inline .github/workflows/draft-release.yml:62]: I think this could be `core.setFailed()` instead.

File: .github/workflows/draft-release.yml [hunk 14]
```diff
@@ -0,0 +1,133 @@
...
+
+            try{
+              const releaseNotesRequest = {
+                owner: context.repo.owner,
+                repo: context.repo.repo,
+                tag_name: sha,
+                target_commitish: sha,
+              };
+
+              if(previousTagName){
```
Comment (sethvargo) [inline .github/workflows/draft-release.yml:76]:
```suggestion
              if(previousTagName) {
```

File: .github/workflows/draft-release.yml [hunk 15]
```diff
@@ -0,0 +1,133 @@
...
+                releaseNotesRequest.previous_tag_name = previousTagName;
+              }
+
+              const releaseNotes = await github.rest.repos.generateReleaseNotes(releaseNotesRequest);
+
+              return releaseNotes.data.body;
+            }catch(err){
```
Comment (sethvargo) [inline .github/workflows/draft-release.yml:83]:
```suggestion
            } catch(err) {
```

File: .github/workflows/draft-release.yml [hunk 16]
```diff
@@ -0,0 +1,133 @@
...
+              core.error(`Failed to generate release notes: ${err}`);
+              process.exit(1);
```
Comment (sethvargo) [inline .github/workflows/draft-release.yml:85]: Same about `setFailed`

File: .github/workflows/draft-release.yml [hunk 17]
```diff
@@ -0,0 +1,133 @@
...
+        with:
+          script: |-
+            const sha = `${{ github.sha }}`;
+            const tag = `v${{ steps.build.outputs.new_version }}`;
+            const baseBranch = `${{ github.ref_name }}`;
+            const repoURL = `${{ github.server_url }}/${{ github.repository }}`;
+            const prBranch = `${{ env.PR_BRANCH }}`;
+            const prBody = `${{ steps.release_notes.outputs.result }}`;
+
+            try{
```
Comment (sethvargo) [inline .github/workflows/draft-release.yml:99]:
```suggestion
            try {
```

File: .github/workflows/draft-release.yml [hunk 18]
```diff
@@ -0,0 +1,133 @@
...
+              const listResponse = await github.rest.pulls.list({
+                owner: context.repo.owner,
+                repo: context.repo.repo,
+                state: `open`,
+                head: prBranch,
+                base: baseBranch,
+              });
+
+              if(!listResponse.data.length){
```
Comment (sethvargo) [inline .github/workflows/draft-release.yml:108]:
```suggestion
              if(!listResponse.data.length) {
```

File: .github/workflows/draft-release.yml [hunk 19]
```diff
@@ -0,0 +1,133 @@
...
+                  owner: context.repo.owner,
+                  repo: context.repo.repo,
+                  title: `Release: ${tag}`,
+                  body: prBody,
+                  head: prBranch,
+                  base: baseBranch,
+                });
+
+                core.info(`Created PR #${createResponse.data.number} at ${createResponse.data.html_url}`);
+              }else{
```
Comment (sethvargo) [inline .github/workflows/draft-release.yml:119]:
```suggestion
              } else {
```

File: .github/workflows/draft-release.yml [hunk 20]
```diff
@@ -0,0 +1,133 @@
...
+                  owner: context.repo.owner,
+                  repo: context.repo.repo,
+                  pull_number: listResponse.data[0].number,
+                  title: `Release: ${tag}`,
+                  body: prBody,
+                });
+
+                core.info(`Updated PR #${updateResponse.data.number} at ${updateResponse.data.html_url}`);
+              }
+            }catch(err){
```
Comment (sethvargo) [inline .github/workflows/draft-release.yml:130]:
```suggestion
            } catch(err) {
```

File: .github/workflows/draft-release.yml [hunk 21]
```diff
@@ -0,0 +1,133 @@
...
+              core.error(`Failed to create pull request: ${err}`);
+              process.exit(1);
```
Comment (sethvargo) [inline .github/workflows/draft-release.yml:132]: Same about `setFailed`

File: .github/workflows/release.yml [hunk 22]
```diff
@@ -0,0 +1,111 @@
...
+  workflow_call:
+
+jobs:
+  release:
+    name: 'Release'
+    runs-on: 'ubuntu-latest'
+    steps:
+      - uses: 'actions/checkout@v3'
+
+      - name: 'Get Package Version'
```
Comment (sethvargo) [inline .github/workflows/release.yml:13]:
```suggestion
      - name: 'Get package version'
```

File: .github/workflows/release.yml [hunk 23]
```diff
@@ -0,0 +1,111 @@
...
+        id: 'version'
+        run: |-
+          echo "::set-output name=new_version::$(cat package.json | jq -r .version)"
+
+      - name: 'Get Previous Release Tag'
```
Comment (sethvargo) [inline .github/workflows/release.yml:18]:
```suggestion
      - name: 'Get previous release tag'
```

File: .prettierrc.js [hunk 24]
```diff
@@ -0,0 +1,13 @@
+module.exports = {
```
Comment (sethvargo) [inline .prettierrc.js:1]: Is this needed?

File: .github/workflows/release.yml [hunk 25]
```diff
@@ -0,0 +1,111 @@
...
+        id: 'prev_release'
+        uses: 'actions/github-script@v5'
+        with:
+          result-encoding: 'string'
+          script: |-
+            const sha = `${{ github.sha }}`;
+
+            try{
```
Comment (sethvargo) [inline .github/workflows/release.yml:26]:
```suggestion
            try {
```

File: .github/workflows/release.yml [hunk 26]
```diff
@@ -0,0 +1,111 @@
...
+              const latestRelease = await github.rest.repos.getLatestRelease({
+                owner: context.repo.owner,
+                repo: context.repo.repo,
+              });
+              return latestRelease.data.tag_name;
+            }catch(err){
```
Comment (sethvargo) [inline .github/workflows/release.yml:32]:
```suggestion
            } catch(err) {
```

File: .github/workflows/release.yml [hunk 27]
```diff
@@ -0,0 +1,111 @@
...
+              if(err['status'] === 404){
```
Comment (sethvargo) [inline .github/workflows/release.yml:33]:
```suggestion
              if(err['status'] === 404) {
```

File: .github/workflows/release.yml [hunk 28]
```diff
@@ -0,0 +1,111 @@
...
+                core.warning(`No releases found`);
+                return '';
+              }else{
```
Comment (sethvargo) [inline .github/workflows/release.yml:36]:
```suggestion
              } else {
```

File: .github/workflows/release.yml [hunk 29]
```diff
@@ -0,0 +1,111 @@
...
+                core.error(`Failed to load latest release: ${err}`);
+                process.exit(1);
+              }
+            }
+
+      - name: 'Create Release'
```
Comment (sethvargo) [inline .github/workflows/release.yml:42]:
```suggestion
      - name: 'Create release'
```

File: .github/workflows/release.yml [hunk 30]
```diff
@@ -0,0 +1,111 @@
...
+        id: 'release'
+        uses: 'actions/github-script@v5'
+        with:
+          script: |-
+            const sha = `${{ github.sha }}`;
+            const tag = `v${{ steps.version.outputs.new_version }}`;
+            const previousTagName = `${{ steps.prev_release.outputs.result }}`;
+
+            try{
```
Comment (sethvargo) [inline .github/workflows/release.yml:51]:
```suggestion
            try {
```

File: .github/workflows/release.yml [hunk 31]
```diff
@@ -0,0 +1,111 @@
...
+              const createReleaseRequest = {
+                owner: context.repo.owner,
+                repo: context.repo.repo,
+                tag_name: tag,
+                target_commitish: sha,
+                name: tag,
+                generate_release_notes: true,
+              };
+
+              if(previousTagName){
```
Comment (sethvargo) [inline .github/workflows/release.yml:61]:
```suggestion
              if(previousTagName) {
```

File: .github/workflows/release.yml [hunk 32]
```diff
@@ -0,0 +1,111 @@
...
+                });
+                
+                createReleaseRequest.body = releaseNotes.data.body;
+                createReleaseRequest.generate_release_notes = false;
+              }
+
+              const response = await github.rest.repos.createRelease(createReleaseRequest);
+
+              core.info(`Created release ${response.data.name} at ${response.data.html_url}`);
+            }catch(err){
```
Comment (sethvargo) [inline .github/workflows/release.yml:77]:
```suggestion
            } catch(err) {
```

File: .github/workflows/release.yml [hunk 33]
```diff
@@ -0,0 +1,111 @@
...
+              core.error(`Failed to create release: ${err}`);
+              process.exit(1);
+            }
+
+      - name: 'Update Floating Tag'
```
Comment (sethvargo) [inline .github/workflows/release.yml:82]:
```suggestion
      - name: 'Update floating tag'
```

File: .github/workflows/draft-release.yml [hunk 34]
```diff
@@ -0,0 +1,133 @@
...
+            }catch(err){
+              if(err['status'] !== 404){
+                core.error(`Failed to load latest release: ${err}`);
+                process.exit(1);
+              }
+
+              core.info(`No existing releases found`);
+            }
+
+            try{
```
Comment (sethvargo) [inline .github/workflows/draft-release.yml:68]:
```suggestion
            try {
```

File: .prettierrc.js (hunk 24 above)
Comment (verbanicm) [reply .prettierrc.js:1]: My default formatter formats YAML with double quotes, so just wanted to ensure that standards.

Comment (verbanicm) [reply .prettierrc.js:1]: Let me know if you want me to remove it

**Review by sethvargo — APPROVED** [review]

**Review by Shasty414 — COMMENTED** [review]
(removed)

**Pull request merged**